be/src/pipeline/task_queue.cpp

```diff
@@ -0,0 +1,192 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "task_queue.h"
+
+namespace doris {
+namespace pipeline {
+
+PipelineTask* SubWorkTaskQueue::try_take() {
+    if (_queue.empty()) {
+        return nullptr;
+    }
+    ++_schedule_time;
+    auto task = _queue.front();
+    _queue.pop();
+    return task;
+}
+
+////////////////////  WorkTaskQueue ////////////////////
+
+WorkTaskQueue::WorkTaskQueue() : _closed(false) {
+    double factor = 1;
+    for (int i = 0; i < SUB_QUEUE_LEVEL; ++i) {
+        _sub_queues[i].set_factor_for_normal(factor);
+        factor *= LEVEL_QUEUE_TIME_FACTOR;
+    }
+
+    int i = 0;
+    _task_schedule_limit[i] = BASE_LIMIT * (i + 1);
+    for (i = 1; i < SUB_QUEUE_LEVEL - 1; ++i) {
+        _task_schedule_limit[i] = _task_schedule_limit[i - 1] + BASE_LIMIT * (i + 1);
+    }
+}
+
+void WorkTaskQueue::close() {
+    std::unique_lock<std::mutex> lock(_work_size_mutex);
+    _closed = true;
+    _wait_task.notify_all();
+}
+
+PipelineTask* WorkTaskQueue::try_take_unprotected() {
+    if (_total_task_size == 0 || _closed) {
+        return nullptr;
+    }
+    double normal_schedule_times[SUB_QUEUE_LEVEL];
+    double min_schedule_time = 0;
+    int idx = -1;
+    for (int i = 0; i < SUB_QUEUE_LEVEL; ++i) {
+        normal_schedule_times[i] = _sub_queues[i].schedule_time_after_normal();
+        if (!_sub_queues[i].empty()) {
+            if (idx == -1 || normal_schedule_times[i] < min_schedule_time) {
+                idx = i;
+                min_schedule_time = normal_schedule_times[i];
+            }
+        }
+    }
+    DCHECK(idx != -1);
```

Comment:
if all `_sub_queue` is empty() ? the DCHECK may failed ?

Reply:
if `_total_task_size == 0`,  `nullptr` will return in the begin of this method.